README.md
# effectR: An R package to call cytoplasmic effectors

[![CRAN version](http://www.r-pkg.org/badges/version/effectR)](https://cran.r-project.org/package=effectR)

# What is *effectR*

The `effectR` package is an R package designed to call oomycete RxLR and CRN effectors by searching for the motifs of interest using regular expression searches and hidden markov models (HMM). 

# What's new in version 1.0.2?

- New test dataset with *P. infestans* reference RxLR effectors 

# Overview

The `effectR` packages searches for the motifs of interest (RxLR-EER motif for RxLR effectors and LFLAK motif for CRN effectors) using a regular expression search (`REGEX`). 
These motifs used by the REGEX `effectR` search have been reported in the literature ([Haas et al., 2009](https://www.nature.com/nature/journal/v461/n7262/full/nature08358.html), [Stam et al., 2013](http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0059517)).

The `effectR` package aligns the REGEX search results using [`MAFFT`](http://mafft.cbrc.jp/alignment/software/), and builds a HMM profile based on the multiple sequence alignment result using the `hmmbuild` program from [`HMMER`](http://hmmer.org/). The HMM profile is used to search across ORF of the genome of interest using the `hmmsearch` binary from `HMMER`. 
The search step will retain sequences with significant hits to the profile of interest. 
`effectR` also combines the redundant sequences found in both REGEX and HMM searches into a single dataset that can be easily exported. 
In addition, `effectR` reads and returns the HMM profile to the user and allows for the creation of a [MOTIF logo-like plot](https://en.wikipedia.org/wiki/Sequence_logo) using `ggplot2`.

## Requirements 

- R packages:
 - [`seqinr`](https://CRAN.R-project.org/package=seqinr/seqinr.pdf)
 - [`ggplot2`](http://ggplot2.org/)
 
- External software 
  - [`MAFFT`](mafft.cbrc.jp/alignment/software/)
  - [`HMMER`](http://hmmer.org/)
 
## Installation

### From CRAN

The latest version of `effectR` can be installed from CRAN. To install, make sure R is at least version 3.4.0. In the R console type

```
install.packages("effectR")
```

### From GitHub

To install `effectR` via GitHub, make sure that the `devtools` package is installed (use `install.packages("devtools")`). After installing devtools, in the R console type:

```
devtools::install_github(repo = "grunwaldlab/effectR", build_vignettes = TRUE)
library("effectR")
```

### External software (REQUIRED)

The `effectR` package uses `MAFFT` and `HMMER3` to perform the hidden markov model seach across the results from the REGEX step. These two packages should be installed before running any of the `effectR` functions.

### Downloading and installing MAFFT

MAFFT is a multiple sequence alignment program that uses Fourier-transform algorithms to align multiple sequences. We recommend downloading and installing MAFFT by following the instructions and steps in the [MAFFT installation](http://mafft.cbrc.jp/alignment/software/) web site. 

#### Linux/OS X Users

Make sure that you remember the directory in which `MAFFT` is installed, of if the installation is sucessful, make sure to obtain the path via bash/tsh/console:

```bash
which mafft
```

```
/usr/local/bin/mafft
```
For more information about MAFFT go to the MAFFT website: http://mafft.cbrc.jp/

#### Windows Users

MAFFT comes in two main distributions for windows:

- [A version that requires a UNIX-like interface called Cygwin](http://mafft.cbrc.jp/alignment/software/windows_cygwin.html)
- [An "all-in-one" version"](http://mafft.cbrc.jp/alignment/software/windows_without_cygwin.html)

Please, download and install the **all-in-one** version.
We recommend that you download and save MAFFT in your Desktop, as it will make yyour path easily accesible.

### Downloading and installing HMMER

HMMER is used for searching sequence databases for sequence homologs. It uses hidden Markov models (profile HMMs) to search for sequences with hits to similar patterns than the profile. We use three main HMMER tools: 

  - `hmmbuild` to create the HMM database from the sequences ontained in the REGEX step of `effectR`
  - `hmmpress` converts the HMM database into a format usable by other `HMMER` programs
  - `hmmsearch` to excecute the HMM search in our sequence queries basde on the HMM profile

The `effectR` package requires all of these tools. A correct `HMMER` installation will install all three programs.

#### Linux/OS X users 

We recommend downloading and installing HMMER by following the instructions and steps in the [HMMER installation](eddylab.org/software/hmmer3/3.1b2/Userguide.pdf) web site. Make sure that you remember the directory in which `HMMER` is installed, of if the installation is sucessful, make sure to obtain the path via bash/tsh/console:

```bash
which hmmbuild
which hmmpress
which hmmsearch
```

```
/usr/local/bin/hmmbuild
/usr/local/bin/hmmpress
/usr/local/bin/hmmsearch
```

For more information about HMMER go to the HMMER website: http://hmmer.org/

#### Windows users

To use the `effectR` package in Windows, the user **must** download the Windows binaries of [HMMER](http://hmmer.org/binaries/hmmer3.0_windows.zip). `effectR` will not work with any other version of HMMER.  
## Data input

The `effectR` package is designed to work with amino acid sequences in [FASTA format](https://en.wikipedia.org/wiki/FASTA_format) representing the six-frame translation of every open reading frame (ORF) of an oomycete genome. 
Using the six-frame translation of all ORF's in a genome is recommended in order to obtain as many effectors as possible from a proteome. 
To obtain the ORF for a genome, we recommend the use of EMBOSS' [`getorf`](http://emboss.sourceforge.net/apps/cvs/emboss/apps/getorf.html).

`effectR` uses a list of sequences of the class `SeqFastadna` in order to perform the effector searches. 
The function `read.fasta` from the `seqinr` package reads the FASTA amino acid file into R, creating a list of `SeqFastadna` objects that represent each of the translated ORF's from the original FASTA file. 

## REGEX search

To perform the effector search, `effectR` searches for the motifs of interest found in RxLR and CRN motifs. 
We have created the function `regex.search` to perform the seach of the motif of interest. 
The function `regex.search` requires the list of `SeqFastadna` objects and the gene family of interest. 

## Multiple sequence alignment and HMMER search

To perform the HMM search and obtain all possible effector candidates from a proteome, `effectR` uses the `REGEX` results as a template to create a HMM profile and perform a search across the proteome of interest.
We have created the `hmm.search` function in order to perfomr this search.
The `hmm.search` function requires a local installation of `MAFFT` and `HMMER` in order to perform the searches. The *absolute paths* of the binaries must be specified in the `mafft.path` and `hmmer.path` options of the `hmm.search` function.
In addition, the `hmm.function` requires the path of the original FASTA file containing the translated ORF's in the `original.seq` parameter of the function. `hmm.search` will use this file as a query in the `hmmsearch` software from HMMER, and search for all sequences with hits against the HMM profile created with the REGEX results.

A default `hmm.search` object returns a list of 3 elements: 

1. The REGEX sequences used to build the HMM profile in a `SeqFastadna` class
2. The sequences from the original translated ORF files with hits to the HMM profile in a `SeqFastadna` class
3. The HMM profile table created by HMMER's `hmmbuild` as a data frame

*NEW FEATURES*:
- `hmm.search` can use a user-defined alignment file (i.e. A multiple sequence alignment performed in MUSCLE, ClustalW, etc.) and omit the alignment step
- `hmm.search` allows the user to save the multiple sequence alignment created by MAFFT within the function

More information on these new features is available in the package help (`?hmm.search`) or in the *effectR* vignette

## Obtaining non-redundant effectors and motif summaries

The user can extract all of the non-redundant sequences and a summary table with the information about the motifs using the `effector.summary` function. 
This function uses the results from either `hmm.seach` or `regex.search` functions to generate a table that includes the name of the candidate effector sequence, the number of motifs of interest (RxLR-EER or LFLAK-HVLV) per sequence and its location within the sequence. 
In addition, when the `effector.summary` function is used in an object that contains the results of `hmm.search`, the user will obtain a list of the non-reduntant sequences. If the user provides the results from `regex.search`, the function will return the motif summary table.

The motif table has a column called **MOTIF**.
This column summarizes the candidate ORF into one of 4 categories:

- Complete: The candidate ORF has both motifs of interest (RxLR + EER or LFLAK + HVLV)
- Only RxLR/Only LFLAK: The candidate ORF only has the translocation domain
- Only EER/HVLV: The candidate ORF only has the second motif of interest
- No motifs: The sequence has a hit with the HMM profile but does not have any motif of interest

### Non-redundant sequences

To export the non-redundant effector candidates that resulted from the `hmm.search` or `regex.search` functions, we use the `write.fasta` function of the `seqinr` package. 
We recomend the users to read the documentation of the [`seqinr`](https://cran.r-project.org/package=seqinr) package
Since the objects that result from the `hmm.search` or `regex.search` function are of the `SeqFastadna` class, we can use any of the function of the `seqinr` package that use this class as well. 

## Visualizing the HMM profile using a sequence logo-like plot

To determine if the HMM profile includes the motifs of interest, we have created the function `hmm.logo`.
The function `hmm.logo` reads the HMM profile (obtained from the `hmm.search` step) and uses `ggplot2` to create a bar-plot.
The bar-plot will illustrate the bits (aminoacid score) of each amino acid used to construct the HMM profile according to its consensus position in the HMM profile. 
To learn more about sequence logo plots visit this [wikipedia article](https://en.wikipedia.org/wiki/Sequence_logo).


## Adding custom motifs to effectR

The effectR package has the capability to use custom regular expressions to predict other families of genes of interest other than RxLR/CRN effector proteins. This example uses the PAAR motif (PAAR) identified in proteins associated with the terminal spike in T6SS of some bacterial species:

```
# Loading the effectR package
library(“effectR”)

# Using the read.fasta function of the seqinr package to import the V. cholerae FASTA proteome file
fasta.file <- " V_cholerae_ATCC_39315.AA.fasta.gz”
ORF <- seqinr::read.fasta(fasta.file)

# Step 1 prediction: Since the PAAR motif can occur anywhere in the sequence of the protein, the REGEX will be very simple and will only contain the PAAR motif.
REGEX <- regex.search(ORF, motif = "custom", reg.pat = "PAAR")
```

Step 1 resulted in a total of 19 predicted proteins with the PAAR motif. We can expand the number of candidate PAAR proteins using the HMM step:

```
# Expanding the search of RxLR effectors using HMM searches (step 2). All candidate effectors predicted by step 2 will be saved in the candidate.rxlr object

candidate.paar <- hmm.search(original.seq = fasta.file, regex.seq = REGEX)

```

Step 2 resulted in one additional candidate protein with a plausible PAAR motif. We can summarize all the information from effectR using the `effector.summary()` function. It will return a table with the candidate proteins, the number of PAAR motifs within each protein, and the position of said motif:

```
# Summarizing the results of effectR.

effector.summary(candidate.paar)

                                  Sequence ID Motif number Motif position        MOTIF
tr|Q9KN60|Q9KN60_VIBCH tr|Q9KN60|Q9KN60_VIBCH            2          35,71 Custom motif
sp|Q9KR02|RUVB_VIBCH     sp|Q9KR02|RUVB_VIBCH            1            145 Custom motif
sp|Q9KPV0|GLND_VIBCH     sp|Q9KPV0|GLND_VIBCH            1            398 Custom motif
sp|Q9KSQ2|HUTG_VIBCH     sp|Q9KSQ2|HUTG_VIBCH            1            269 Custom motif
sp|Q9KPU5|NUSB_VIBCH     sp|Q9KPU5|NUSB_VIBCH            1              7 Custom motif
tr|Q9KS85|Q9KS85_VIBCH tr|Q9KS85|Q9KS85_VIBCH            1            171 Custom motif
tr|Q9KUC8|Q9KUC8_VIBCH tr|Q9KUC8|Q9KUC8_VIBCH            1            232 Custom motif
tr|Q9KND6|Q9KND6_VIBCH tr|Q9KND6|Q9KND6_VIBCH            1            153 Custom motif
tr|Q9KN94|Q9KN94_VIBCH tr|Q9KN94|Q9KN94_VIBCH            1            236 Custom motif
tr|Q9KMP0|Q9KMP0_VIBCH tr|Q9KMP0|Q9KMP0_VIBCH            1             35 Custom motif
tr|Q9KPU1|Q9KPU1_VIBCH tr|Q9KPU1|Q9KPU1_VIBCH            1            179 Custom motif
tr|Q9KSK0|Q9KSK0_VIBCH tr|Q9KSK0|Q9KSK0_VIBCH            1            501 Custom motif
tr|Q9KLU5|Q9KLU5_VIBCH tr|Q9KLU5|Q9KLU5_VIBCH            1            481 Custom motif
tr|Q9KKR8|Q9KKR8_VIBCH tr|Q9KKR8|Q9KKR8_VIBCH            1            343 Custom motif
tr|Q9KPP4|Q9KPP4_VIBCH tr|Q9KPP4|Q9KPP4_VIBCH            1            811 Custom motif
tr|Q9KUF6|Q9KUF6_VIBCH tr|Q9KUF6|Q9KUF6_VIBCH            1            290 Custom motif
tr|Q9KVN5|Q9KVN5_VIBCH tr|Q9KVN5|Q9KVN5_VIBCH            1             74 Custom motif
tr|Q9KQJ5|Q9KQJ5_VIBCH tr|Q9KQJ5|Q9KQJ5_VIBCH            1            189 Custom motif
tr|Q9KNT2|Q9KNT2_VIBCH tr|Q9KNT2|Q9KNT2_VIBCH            1            146 Custom motif
tr|Q9KUM6|Q9KUM6_VIBCH tr|Q9KUM6|Q9KUM6_VIBCH            0           <NA>    No MOTIFS
```

The results illustrate that 19 out of the 20 candidate proteins have a predicted PAAR domain within its sequence, and only protein `tr|Q9KN60|Q9KN60_VIBCH` has more than 1 PAAR motif. 

Any user can add the motif of interest into the effectR package by adding a simple line of code within the `regex.search` function. We will illustrate this feature by adding the PAAR motif search as part of the `regex.search` function:

```
# The regex.search function
regex.search <- function(sequence, motif = "RxLR", reg.pat = NULL){
 if (unique(unlist(lapply(sequence, class))) != "SeqFastadna") {
    stop("The object is not a list of sequences read by seqinr.")
  }
  seq <- lapply(sequence, function (x) paste(unlist(x),collapse = ""))
    regex <- list()
    if (motif %in% c("RxLR","CRN",”PAAR”) & !is.null(reg.pat)){
      message(paste0("Custom REGEX patterns are not supported with the 'CRN' or 'RxLR' motif options.\n The package will use the default REGEX patterns used to search for ", motif, " motifs."))
      Sys.sleep(2)
    }
    for (i in 1:length(seq)){
      if (motif == "RxLR"){
        reg.pat <- "^\\w{10,40}\\w{1,96}R\\wLR\\w{1,40}eer"
      } else if (motif == "CRN"){
        reg.pat <- "^\\w{1,90}LFLAK\\w+"
      } else if (motif == "PAAR"){
        reg.pat <- “PAAR”
	} else if (motif == "custom"){
        if (is.null(reg.pat)){
          stop("No custom REGEX pattern found.\n The 'custom' option requires a mandatory REGEX pattern")
        } else {
        reg.pat <- reg.pat
        }
      }
      regex[[i]] <- unlist(gregexpr(seq[[i]], pattern = reg.pat, perl = T ,ignore.case = T))
    #percentage <- percentage + 1/length(seq)*100
    }
    regex <- as.data.frame(do.call(rbind, regex))
    regex$seq <- names(seq)
    regex <- regex[!regex$V1 < 0, ]
    regex <- sequence[seqinr::getName(sequence) %in% regex$seq]
    if (length(regex) == 0){
      stop(paste0("No ",motif, " sequences found."))
    }
    return(regex)
}
```

After including the new PAAR motif, the user can specify the PAAR motif as an option of the `motif` parameter:

```
# Loading the effectR package in R
library(“effectR”)

# Using the read.fasta function of the seqinr package to import the V. cholerae FASTA proteome file

fasta.file <- " V_cholerae_ATCC_39315.AA.fasta.gz”
ORF <- seqinr::read.fasta(fasta.file)

# Step 1 prediction: Predict proteins with the PAAR motif

REGEX <- regex.search(ORF, motif = "PAAR)

```

This customization will allow users to add any motif of interest to the *effectR* package by forking the github repository, adding the additional line into the regex.search function, adding the respective reference of the motif into the `@references` section of the R documentation within the `regex.search` function, and submitting a pull request to the package maintainers. The package maintainers will update the package, test the motif, and, if valid, add the motif to the effector.summary function before updating *effectR*. The currently available CRAN version of *effectR* only includes the RxLR and CRN motifs to facilitate the familiarization and engagement of the community with the package, but additional custom REGEX patterns will be added as the package is updated.



NEWS.md
# effectR 1.0.3 (2018-08-02)

# What's new in version 1.0.3?

- User can provide a bit-score threshold for `hmm.search`


inst/shiny/server.R
# Load dependencies
require(shiny)
require(shinyjs)
require(seqinr)
require(ggplot2)
require(viridis)
require(reshape2)

# Max size limit:
options(shiny.maxRequestSize=70*4024^2)

# Code for refresh button
`%then%` <- shiny:::`%OR%`
jscode <- "shinyjs.refresh = function() { history.go(0); }"


### Additional scripts

get_mafft_path <- function(mafft.path = NULL, error = TRUE,
                           verbose = FALSE) {
  # Set default path
  if (is.null(mafft.path)) {
    path <- unname(Sys.which("mafft"))
  } else if (endsWith(mafft.path, "mafft")) {
    path <- mafft.path
  } else if (Sys.info()[['sysname']] %in% "Windows") {
    path <- file.path(mafft.path,"mafft.bat")
  }  else {
    # add on executable to path if not already present
    path <- file.path(mafft.path, "mafft")
  }


  # Check if mafft is installed
  is_installed <- system2(path, "--version", stderr = NULL) == 0
  if (! is_installed && error) {
    if (is.null(mafft.path)) {
      stop(paste0("MAFFT not found in your computer's search path.",
                  "'\n Please check that MAFFT is installed and in the search path or specify the path to the MAFFT installation using the `mafft.path` option."), call. = FALSE)
    } else {
      stop(paste0("MAFFT not found in the specified path: '", path,
                  "'\n Please check your MAFFT installation."), call. = FALSE)
    }
  }

  return(path)
}

# Get HMMER path
get_hmmer_path <- function(command, hmmer.path = NULL, error = TRUE,
                           verbose = FALSE) {
  # Set defualt path
  if (is.null(hmmer.path)) {
    path <- unname(Sys.which(command))
  } else {
    path <- file.path(hmmer.path, command)
  }

  # Print progress
  if (verbose) {
    message(paste0("Checking if HMMER is installed in the specified path: '", path,"'"))
  }

  # Check if mafft is installed
  is_installed <- system2(path, "-h", stderr = NULL, stdout = NULL) == 0
  if (! is_installed && error) {
    if (is.null(hmmer.path)) {
      stop(paste0("HMMER not found in your computer's search path.",
                  "'\n Please check that HMMER is installed and in the search path or specify the path to the HMMER installation using the `hmm.path` option."), call. = FALSE)
    } else {
      stop(paste0("HMMER not found in the specified path: '", path,
                  "'\n Please check your HMMER installation."), call. = FALSE)
    }
  }

  return(path)
}



# Converts FASTA to STOCKHOLM
fasta_to_stockholm <- function(fasta.file){
  stock.name <- gsub(fasta.file, pattern = ".fasta", replacement = ".stockholm"
  )
  seq <- seqinr::read.fasta(fasta.file)
  seq.seq <- lapply(seqinr::getSequence(seq), function (x) (paste0(x,collapse="")))
  seq.names <- seqinr::getName(seq)
  seq.final <- list()
  for (i in 1:length(seq)){
    seq.final[[i]] <- paste(seq.names[[i]],seq.seq[[i]], sep = " ")
  }
  seq.final <- unlist(seq.final)
  writeLines(c("# STOCKHOLM 1.0", seq.final,"//"), con = stock.name, sep = "\n")
}

mafft.path <- dirname(get_mafft_path())
hmm.path <- dirname(get_hmmer_path("hmmsearch"))
mafft.path.shiny <- mafft.path
hmm.path.shiny <- hmm.path


# All variable names
original.seq <- c("AA.fasta")
file.name <- c("REGEX.fasta")
mafft.out.name <- c("MAFFT.fasta")
hmmbuild.out <- c("hmmbuild.hmm")
hmmsearch.out <- c("hmmsearch.txt")
mafft.path <- mafft.path.shiny
hmm.path <- hmm.path.shiny
original.dir <- getwd()

# TMP dir
tmp.dir <- tempdir()
setwd(tmp.dir)
##

##### SHINY APP START #####
shinyServer(function(input, output, session) {

  # File read in
  fasta <- reactive({
    file <- input$file1
    write.fasta(sequences = read.fasta(file$datapath), names = names(read.fasta(file$datapath)), file.out = "AA.fasta", open = "w", nbchar = 10000)
    read.fasta(file$datapath)
  })

  # UI output
  output$input_fasta <- renderPrint({
    file <- input$file1
    if(is.null(file))
      return("No FASTA file")
    else
    return (paste(length(fasta()), "Sequences read."))
  })

  # Refresh app button
  observeEvent(input$refresh, {
    js$refresh();
  })

########### Pipeline funtions ############

  #####
  # Step 1: REGEX
  ####

  # REGEX Script
  percentage <- 0
  regex.seq <- reactive({
    validate(
      need(is.null(file) == F, "No sequence file loaded") %then%
      need(file.exists("AA.fasta") != F, "No AA file found. Reload the web-app and try again")
    )
    seq <- lapply(fasta(), function (x) paste(unlist(x),collapse = ""))
    withProgress(message = "Searching for sequences that match the REGEX of interest", value=0, {
      regex <- list()
      for (i in 1:length(seq)){
        if (input$motif_sel == "RxLR"){
          regex[[i]] <- unlist(gregexpr(seq[[i]], pattern="^\\w{10,40}\\w{1,96}R\\wLR\\w{1,40}eer", perl = T,ignore.case = T))
        } else if (input$motif_sel == "CRN"){
          regex[[i]] <- unlist(gregexpr(seq[[i]], pattern="^\\w{1,90}LFLAK\\w+", perl = T,ignore.case = T))
        }
        percentage <- percentage + 1/length(seq)*100
        incProgress(1/length(seq), detail = paste0("Progress: ",round(percentage,2),"%"))
      }
      regex <- as.data.frame(do.call(rbind, regex))
      regex$seq <- names(seq)
    })
    write.fasta(sequences = fasta()[names(fasta()) %in% regex[!regex$V1 %in% -1, 2]], names = names(fasta()[names(fasta()) %in% regex[!regex$V1 %in% -1, 2]]), file.out = file.name, open = "w", nbchar = 10000)
    fasta()[names(fasta()) %in% regex[!regex$V1 %in% -1, 2]]
  })

  # REGEX UI interface
  observeEvent(input$regex_seach, {
    output$regex <- renderPrint({
      regex.seq()
      cat("Number of sequences retained by REGEX: ")
      cat(length(regex.seq()))
      cat("\n")
      session$sendCustomMessage("regex_ready", list(fileSize=floor(runif(1) * 10000)))

    })
  })

  output$downloadREGEX <- downloadHandler(
    filename <- function() {
      paste("REGEX", "fasta", sep=".")
    },
    content <- function(file) {
       file.copy("REGEX.fasta", file)
    }
  )


  #####
  # Step 2: HMM search
  ####

    # Script
  observeEvent(input$do, {
    # Step 2A:
    # MAFFT alignment
    output$mafft <- renderPrint({
      validate(
        need(length(regex.seq()) > 4, "Not enough sequences for HMM step. More than 4 sequences with REGEX required to perform HMM search")
      )
      withProgress(message = 'Performing MAFFT alignment', value = 100, {
                mafft.command <- c(get_mafft_path(mafft.path),
                           "--legacygappenalty",
                           "--genafpair",
                           "--maxiterate", "1000",
                           "--quiet",
                           file.name)
        system2(mafft.command, stdout = mafft.out.name )
      })
      cat("MAFFT aligment: Done!")
      cat("\n")
    })

    # Step 2B:
    # HMM build
    output$hmmer_press <- renderPrint({
      validate(
        need(file.exists(mafft.out.name) != F, "No HMM performed")
      )
      withProgress(message = 'Building HMM model...', value = 100, {
        unlink(file.path(tmp.dir, hmmbuild.out))
        if (Sys.info()[['sysname']] %in% "Windows"){
          fasta_to_stockholm(fasta.file = mafft.out.name)
          stock.name <- gsub(mafft.out.name, pattern = ".fasta", replacement = ".stockholm"
          )
          hmmbuild_command <- c(get_hmmer_path("hmmbuild.exe", hmm.path),
                                "--seed 12345",
                                "--amino",
                                hmmbuild.out,
                                stock.name)
        } else {
          hmmbuild_command <- c(get_hmmer_path("hmmbuild", hmm.path),
                                "--seed 12345",
                                "--amino",
                                hmmbuild.out,
                                mafft.out.name)
        }
        system2(hmmbuild_command, stdout = F)
        Sys.sleep(0.2)
        if (Sys.info()[['sysname']] %in% "Windows"){
          hmmpress_command <- c(get_hmmer_path("hmmpress.exe", hmm.path),
                                "-f",
                                hmmbuild.out)
        } else {
          hmmpress_command <- c(get_hmmer_path("hmmpress", hmm.path),
                                "-f",
                                hmmbuild.out)
        }
      system2(hmmpress_command)
      })
      cat("HMM model built!")
      cat("\n")
    })


    # Step 2C:
    # HMM search
    output$hmmer_search <- renderPrint({
      validate(
        need(file.exists("hmmbuild.hmm") != F, "No HMM performed")
      )
      withProgress(message = 'Performing HMM search...', value = 100, {
        if (Sys.info()[['sysname']] %in% "Windows"){
          hmmsearch_command <- c(get_hmmer_path("hmmsearch.exe", hmm.path),
                                 "-T", "0",
                                 "--seed 12345",
                                 "--tblout", hmmsearch.out,
                                 hmmbuild.out,
                                 original.seq)
        } else {
          hmmsearch_command <- c(get_hmmer_path("hmmsearch", hmm.path),
                                 "-T", "0",
                                 "--seed 12345",
                                 "--tblout", hmmsearch.out,
                                 hmmbuild.out,
                                 original.seq)
        }
        system2(hmmsearch_command)
      })
      Sys.sleep(0.2)
      cat("HMM search: Done!")
      cat("\n")
      session$sendCustomMessage("download_ready", list(fileSize=floor(runif(1) * 10000)))
     })
  })

  output$downloadHMM <- downloadHandler(
    filename <- function() {
      paste("HMM_files", "zip", sep=".")
    },
    content <- function(file) {
      fileConn<-file("Readme.txt")
      writeLines(c("Files used in HMM step:", "MAFFT.fasta: MAFFT alignment of REGEX candidates.","hmmbuild.hmm: HMM model of REGEX RxLR candidates", "hmmsearch.txt: Results of HMM search using input file"), fileConn)
      close(fileConn)
      system("zip HMM_data.zip MAFFT.fasta hmmsearch.txt hmmbuild.hmm Readme.txt")
      file.copy("HMM_data.zip", file)
    },
    contentType = "application/zip"
  )


  # OPTIONAL STEP: Web Logo
  observeEvent(input$logo, {
        validate(
          need(length(regex.seq()) > 4, "Not enough sequences to construct the HMM LOGO")
        )
        output$preImage <- renderPlot({
          withProgress(message = 'Constructing HMM logo...', value = 100, {
            # Plot
            if (Sys.info()[['sysname']] %in% "Windows"){
              hmm <- utils::read.table(hmmbuild.out, blank.lines.skip = T, skip = 14, sep = "", fill = T, stringsAsFactors = F)
            } else {
              hmm <- utils::read.table(hmmbuild.out, blank.lines.skip = T, skip = 16, sep = "", fill = T, stringsAsFactors = F)
            }
            colnames(hmm) <- hmm[1,]
            hmm <- hmm[-(1:5),]
            hmm[,1] <- as.numeric(hmm[,1])
            hmm <- hmm[hmm[,1]%%1==0, ]
            hmm <- hmm[!is.na(as.numeric(hmm[,2])), ]
            rownames(hmm) <- hmm[,1]
            hmm <- hmm[,-1]
            hmm <- data.frame(sapply(hmm, function(x) as.numeric(as.character(x))))
            hmm.sums <- apply(hmm,2,function (x) max(x)/x)
            hmm.sums <- apply(hmm.sums,2,function (x) x/sum(x))
            hmm.sums <- apply(hmm.sums, 2, function (x) x - mean(as.numeric(as.character(x))))
            hmm.sums[hmm.sums < 0] <- 0
            # Melt HMM
            hmm.m <- reshape2::melt(t(hmm.sums))
            colnames(hmm.m) <- c("element","position","bits")
            hmm.m$bits <- as.numeric(as.character(hmm.m$bits))
            p <- ggplot(hmm.m, aes(x=position, y=bits, fill=element)) + geom_bar(stat = "identity",position = "stack",width=1, alpha=0.5) + geom_text(aes(label=element, size=bits), position='stack') +  scale_fill_viridis(discrete=TRUE) + theme_bw() + ylab("Relative Frequency (bits)") + guides(fill=FALSE)
            print(p)
          })
          session$sendCustomMessage("logo_ready", list(fileSize=floor(runif(1) * 10000)))
        })
       })


  #####
  # Step 3: Additional steps
  ####

  # Step 3A: Validate and create final dataset
  final_fasta <- reactive({
    regex <- read.fasta("REGEX.fasta")
    if(file.exists("hmmsearch.txt")){
      hmm <- read.table("hmmsearch.txt", stringsAsFactors = F)
      combined.names <- unique(c(hmm$V1, names(regex)))
      fasta()[names(fasta()) %in% combined.names]
    } else {
      fasta()[names(fasta()) %in% names(regex)]
    }
  })

  # Step 3B: Combine datasets and create REGEX table
  # Table
  summary_table <- reactive (
    if(file.exists("hmmsearch.txt")){
    hmm <- read.delim("hmmsearch.txt", comment.char = "#", header = F, sep = "", blank.lines.skip = T, stringsAsFactors = F)[,1]
    num_hmm <- length(hmm)
    num_len <- length(regex.seq())
    combined.names <- unique(c(hmm, names(regex.seq())))
    num_com <- length(combined.names)
    final.df <- data.frame(num_len,num_hmm,num_com)
    colnames(final.df) <- c("REGEX","HMM","Total")
    final.df
  } else {
    num_len <- length(final_fasta())
    final.df <- data.frame(num_len)
    colnames(final.df) <- c("REGEX")
    final.df
  }
  )


  observeEvent(input$combine, {
    output$final_table <- renderTable({
      summary_table()
    })
  })
  # Step 3C: Motif Table
  motif_table <- reactive({
    sequences <- lapply(final_fasta(), function (x) paste(unlist(x),collapse = ""))
    # REGEX search of the motifs of interest:
    if (input$motif_sel == "RxLR"){
      rxlr.num <- unlist(lapply(lapply(sequences, function (x) unlist(gregexpr(x, pattern="r\\wlr", perl = T,ignore.case = T))), function (x) length(x)))
      rxlr.motif <- unlist(lapply(lapply(sequences, function (x) unlist(gregexpr(x, pattern="r\\wlr", perl = T,ignore.case = T))), function (x) paste(x,collapse = ",")))
      eer.num <- unlist(lapply(lapply(sequences, function (x) unlist(gregexpr(x, pattern="[ED][ED][KR]", perl = T,ignore.case = T))), function (x) length(x)))
      eer.motif <- unlist(lapply(lapply(sequences, function (x) unlist(gregexpr(x, pattern="[ED][ED][KR]", perl = T,ignore.case = T))), function (x) paste(x,collapse = ",")))
      motifs <- data.frame(seqinr::getName(final_fasta()),rxlr.num,rxlr.motif,eer.num,eer.motif, stringsAsFactors = F)
    } else if (input$motif_sel == "CRN") {
      lflak.num <- unlist(lapply(lapply(sequences, function (x) unlist(gregexpr(x, pattern="lflak", perl = T,ignore.case = T))), function (x) length(x)))
      lflak.motif <- unlist(lapply(lapply(sequences, function (x) unlist(gregexpr(x, pattern="lflak", perl = T,ignore.case = T))), function (x) paste(x,collapse = ",")))
      hvlv.num <- unlist(lapply(lapply(sequences, function (x) unlist(gregexpr(x, pattern="hvlv", perl = T,ignore.case = T))), function (x) length(x)))
      hvlv.motif <- unlist(lapply(lapply(sequences, function (x) unlist(gregexpr(x, pattern="hvlv", perl = T,ignore.case = T))), function (x) paste(x,collapse = ",")))
      motifs <- data.frame(seqinr::getName(final_fasta()),lflak.num,lflak.motif,hvlv.num,hvlv.motif, stringsAsFactors = F)
    }
    motifs[motifs == -1] <- NA
    motifs <- data.frame(apply(motifs, 2, as.character), stringsAsFactors = F)
    motifs[,2][is.na(motifs[,3])] <- 0
    motifs[,4][is.na(motifs[,5])] <- 0
    motifs <- motifs[order(motifs[,2],motifs[,4],decreasing = T),]
    motifs$summary <- "No MOTIFS"
    motifs$summary[motifs[,2] > 0 & motifs[,4] > 0] <- "Complete"
    motifs$summary[motifs[,2] != 0 & motifs[,4] == 0] <- if (input$motif_sel == "RxLR"){c("Only RxLR motif")} else if (input$motif_sel == "CRN"){("Only LFLAK motif")}
    motifs$summary[motifs[,2] == 0 & motifs[,4] != 0] <- if (input$motif_sel == "RxLR"){c("Only EER motif")} else if (input$motif_sel == "CRN"){("Only HVLV motif")}
    if (input$motif_sel == "RxLR"){
      colnames(motifs) <- c("Sequence ID","RxLR number","RxLR position","EER number","EER position","MOTIF")
    } else if (input$motif_sel == "CRN") {
      colnames(motifs) <- c("Sequence ID","LFLAK number","LFLAK position","HVLV number","HVLV position","MOTIF")
    }
    motifs
  })

  # Render motif table
  observeEvent(input$motif, {
    output$motif_table <- renderTable({
      motif_table()
     })
    session$sendCustomMessage("motif_ready", list(fileSize=floor(runif(1) * 10000)))
  })

  # Download MOTIF table
  output$motifDownload <- downloadHandler(
    # This function returns a string which tells the client
    # browser what name to use when saving the file.
    filename = paste("Motif_table","txt",sep = "."),
    # This function should write data to a file given to it by
    # the argument 'file'.
    content = function(file) {
      # Write to a file specified by the 'file' argument
      write.table(x = motif_table(), file, sep = "\t", quote = F, row.names = F)
    })

  # MOTIF summary table
  observeEvent(input$motif_summ, {
    output$motif_summ_table <- renderTable({
      table(motif_table()$MOTIF)
      })
  })

  # Downloading final RXLR file
  output$downloadData <- downloadHandler(
    # This function returns a string which tells the client
    # browser what name to use when saving the file.
    filename = paste("Candidate_effectors","fasta",sep = "."),
    # This function should write data to a file given to it by
    # the argument 'file'.
    content = function(file) {
      # Write to a file specified by the 'file' argument
    write.fasta(sequences = final_fasta(), names = names(final_fasta()), nbchar = 10000, file.out = file)
    })
})

######### SHINY APP ENDS ##########
# EOF
